builder/steps/gen-dockerfile/contents/src/detect_setup.ts

```diff
@@ -0,0 +1,167 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as yaml from 'js-yaml';
+
+import {Locator, Reader} from './fsview';
+import {Logger} from './logger';
+
+const APP_YAML = 'app.yaml';
+const PACKAGE_JSON = 'package.json';
+const YARN_LOCK = 'yarn.lock';
+
+/**
+ * Encapsulates the information about the Node.js application detected by
+ * the {@link detectSetup} method.
+ */
+export interface Setup {
+  /**
+   * Specifies whether the app directory contains the files necessary to
+   * install dependencies using either `npm install` or `yarn install`
+   */
+  canInstallDeps: boolean;
+  /**
+   * Specifies whether the app directory's package.json file contains a
+   * "scripts" section that contains a "start" command
+   */
+  gotScriptsStart: boolean;
+  /**
+   * Specifies the version of Node.js used to run the application as specified
+   * by the application's package.json file
+   */
+  nodeVersion: string|undefined;
+  /**
+   * Specifies whether or not Yarn should be used to install the application's
+   * dependencies and launch the app.
+   */
+  useYarn: boolean;
+}
+
+/**
+ * This function is used to detect information about a Node.js application.
+ * In particular, this function determines whether or not a `package.json`
+ * file is available to allow `npm` or `yarn` to do an install, whether
+ * the package.json file has a start script specified, whether the package.json
+ * has an engines field specified that specifies a Node.js version to use,
+ * and whether `yarn` or `npm` should be used to install dependencies and run
+ * the application.
+ *
+ * @param logger Used to print statements designed for user consumption.
+ * @param fsview Represents a view of the root directory of a Node.js
+ *               application.
+ *
+ * @throws If an `app.yaml` file does not exist in the directory desribed by
+ *         the given {@link Locator}.
+ * @throws If a `package.json` file exists but a problem occurred while
+ *         trying to parse the file.
+ * @throws If the directory described by the {@link Reader} and {@link Locator}
+ *         does not have a start script in the `package.json` file (or the
+ *         `package.json` file doesn't exist) and the directory does not
+ *         contain a `server.js` file.
+ *
+ * @return Information about the Node.js application in question that can
+ *         be used to determine how to install the application's dependencies
+ *         and run the application.
+ */
+export async function detectSetup(logger: Logger, fsview: Reader&Locator): Promise<Setup> {
+  if (!(await fsview.exists(APP_YAML))) {
```

Comment:
I know a race condition can occur, but for this particular case it won't be an issue because of how the code will be used. Also, `fsview.exists` uses `fs.existsSync`, which isn't deprecated.  These are things I want to address but aren't as high on my list of things todo, and could probably be done on a follow-up.